File: web/src/core/matcher.ts
import { Board } from "./board";
import { canSwap, GameConfig, GridPos, isMatchable, PieceType, gridPos } from "./types";

export class Matcher {
  constructor(
    private readonly board: Board,
    private readonly config: GameConfig,
  ) {}

  findAllMatches(): Set<GridPos> {
    const matches = new Set<string>();
    const result = new Set<GridPos>();

    const add = (positions: GridPos[]) => {
      for (const pos of positions) {
        const key = `${pos.x},${pos.y}`;
        if (!matches.has(key)) {
          matches.add(key);
          result.add(pos);
        }
      }
    };

    for (let y = 0; y < this.board.height; y++) {
      for (let x = 0; x < this.board.width; x++) {
        add(this.getHorizontalMatch(x, y));
      }
    }

    for (let x = 0; x < this.board.width; x++) {
      for (let y = 0; y < this.board.height; y++) {
        add(this.getVerticalMatch(x, y));
      }
    }

    return result;
  }

  hasValidMoves(): boolean {
    for (let x = 0; x < this.board.width; x++) {
      for (let y = 0; y < this.board.height; y++) {
        const pos = gridPos(x, y);
        if (!canSwap(this.board.getPiece(pos))) continue;

        if (x + 1 < this.board.width) {
          const right = gridPos(x + 1, y);
          if (canSwap(this.board.getPiece(right)) && this.wouldMatchAfterSwap(pos, right)) {
            return true;
          }
        }

        if (y + 1 < this.board.height) {
          const down = gridPos(x, y + 1);
          if (canSwap(this.board.getPiece(down)) && this.wouldMatchAfterSwap(pos, down)) {
            return true;
          }
        }
      }
    }
    return false;
  }

  private wouldMatchAfterSwap(a: GridPos, b: GridPos): boolean {
    this.board.swap(a, b);
    const hasMatch = this.findAllMatches().size > 0;
    this.board.swap(a, b);
    return hasMatch;
  }

  private getHorizontalMatch(x: number, y: number): GridPos[] {
    const start = gridPos(x, y);
    const type = this.board.getPiece(start);
    if (!isMatchable(type)) return [];

    const result: GridPos[] = [start];
    for (let i = x + 1; i < this.board.width; i++) {
      if (this.board.getPiece(gridPos(i, y)) === type) {
        result.push(gridPos(i, y));
      } else {
        break;
      }
    }

    return result.length >= this.config.minMatchCount ? result : [];
  }

  private getVerticalMatch(x: number, y: number): GridPos[] {
    const start = gridPos(x, y);
    const type = this.board.getPiece(start);
    if (!isMatchable(type)) return [];

    const result: GridPos[] = [start];
    for (let i = y + 1; i < this.board.height; i++) {
      if (this.board.getPiece(gridPos(x, i)) === type) {
        result.push(gridPos(x, i));
      } else {
        break;
      }
    }

    return result.length >= this.config.minMatchCount ? result : [];
  }
}

File: Assets/Scripts/Gameplay/Matcher.cs
using System.Collections.Generic;
using Match3.Data;

namespace Match3.Gameplay
{
    /// <summary>
    /// 匹配检测：符文/炸弹可匹配，冰块不参与颜色匹配。
    /// </summary>
    public class Matcher
    {
        private readonly Board _board;
        private readonly GameConfig _config;

        public Matcher(Board board, GameConfig config)
        {
            _board = board;
            _config = config;
        }

        public HashSet<GridPos> FindAllMatches()
        {
            var matches = new HashSet<GridPos>();

            for (var y = 0; y < _board.Height; y++)
            {
                for (var x = 0; x < _board.Width; x++)
                {
                    foreach (var pos in GetHorizontalMatch(x, y))
                        matches.Add(pos);
                }
            }

            for (var x = 0; x < _board.Width; x++)
            {
                for (var y = 0; y < _board.Height; y++)
                {
                    foreach (var pos in GetVerticalMatch(x, y))
                        matches.Add(pos);
                }
            }

            return matches;
        }

        public bool HasValidMoves()
        {
            for (var x = 0; x < _board.Width; x++)
            {
                for (var y = 0; y < _board.Height; y++)
                {
                    var pos = new GridPos(x, y);
                    if (!PieceTypeHelper.CanSwap(_board.GetPiece(pos)))
                        continue;

                    if (x + 1 < _board.Width)
                    {
                        var right = new GridPos(x + 1, y);
                        if (PieceTypeHelper.CanSwap(_board.GetPiece(right)) &&
                            WouldMatchAfterSwap(pos, right))
                            return true;
                    }

                    if (y + 1 < _board.Height)
                    {
                        var down = new GridPos(x, y + 1);
                        if (PieceTypeHelper.CanSwap(_board.GetPiece(down)) &&
                            WouldMatchAfterSwap(pos, down))
                            return true;
                    }
                }
            }

            return false;
        }

        private bool WouldMatchAfterSwap(GridPos a, GridPos b)
        {
            _board.Swap(a, b);
            var hasMatch = FindAllMatches().Count > 0;
            _board.Swap(a, b);
            return hasMatch;
        }

        private List<GridPos> GetHorizontalMatch(int x, int y)
        {
            var result = new List<GridPos> { new GridPos(x, y) };
            var type = _board.GetPiece(new GridPos(x, y));
            if (!PieceTypeHelper.IsMatchable(type))
                return new List<GridPos>();

            for (var i = x + 1; i < _board.Width; i++)
            {
                if (_board.GetPiece(new GridPos(i, y)) == type)
                    result.Add(new GridPos(i, y));
                else
                    break;
            }

            return result.Count >= _config.MinMatchCount ? result : new List<GridPos>();
        }

        private List<GridPos> GetVerticalMatch(int x, int y)
        {
            var result = new List<GridPos> { new GridPos(x, y) };
            var type = _board.GetPiece(new GridPos(x, y));
            if (!PieceTypeHelper.IsMatchable(type))
                return new List<GridPos>();

            for (var i = y + 1; i < _board.Height; i++)
            {
                if (_board.GetPiece(new GridPos(x, i)) == type)
                    result.Add(new GridPos(x, i));
                else
                    break;
            }

            return result.Count >= _config.MinMatchCount ? result : new List<GridPos>();
        }
    }
}
